docs/performance/inserts/bulk.txt

```diff
@@ -0,0 +1,147 @@
+easily handle several thousand inserts per second.
+
+However, it's important to note that:
+
+ 1. Every insert is first applied to the primary shard
+
+ 2. After the primary shard has been updated, the insert is then individually
+    communicated in parallel to every configured replica shard
+
+ 3. CrateDB will not return a response until all replica shards have been
+    updated
+
+The overhead for each insert (query parsing, planning, and execution, which
+includes network connection setups, connection teardowns, and connection
+metadata) starts to add up for very heavy insert workloads.
+
+In addition, lots of internal traffic will congest your network, which will
+slow down any network-based cluster operations (i.e. inserts, distributed
+reads, replication, and cluster management).
+
+Fortunately, there is a solution: bulk inserts.
+
+Switching to Bulk Inserts
+=========================
+
+If you can batch up your inserts, switching to *bulk inserts* will dramatically
+improve both ingestion performance, and overall cluster performance.
+
+Bulk inserts make use of the CrateDB HTTP endpoint `bulk operations`_ feature
+to perform many inserts in a single operation.
+
+Bulk inserts still generate internal network traffic, and CrateDB still waits
+until all replicas have been updated before returning a response. But there is
+less internal network traffic. And in addition, the bulk query only needs to be
+parsed, planned, and executed once.
+
+Using ``UNNEST`` As a Fallback
+==============================
+
+If the HTTP endpoint isn't a viable option for you, CrateDB also provides
+support for bulk inserts via the `UNNEST`_ function. This is a less elegant way
+of doing bulk inserts. But you can expect to see similar performance
+improvements.
+
+The ``UNNEST`` function produce rows, like so:
+
+.. code-block:: psql
+
+    cr> SELECT *
+    ...   FROM UNNEST(
+    ...          [1, 2, 3],
+    ...          ['Arthur', 'Trillian', 'Marvin']);
+    +------+----------+
+    | col1 | col2     |
+    +------+----------+
+    |    1 | Arthur   |
+    |    2 | Trillian |
+    |    3 | Marvin   |
+    +------+----------+
+    SELECT 3 rows in set (... sec)
+
+Combine ``UNNEST`` with ``INSERT`` to perform bulk inserts:
+
+.. code-block:: psql
+
+    cr> INSERT INTO my_table (id, name)
+    ...      VALUES (SELECT *
+    ...                FROM UNNEST(
+    ...                       [1, 2, 3],
+    ...                       ['Arthur', 'Trillian', 'Marvin']));
+    INSERT OK, 3 rows affected (... sec)
+
+Performance Testing
+===================
+
+Every setup is different. The best way to tune your particular setup is to do
+performance testing.
+
+The easiest way to do this is to use the `cr8`_ tool.
+
+Let's say you have CrateDB running on ``localhost:4200``, and you want to test
+inserts against the ``my_table`` table. You can run cr8, like so:
+
+.. code-block:: sh
+
+    cr> time cr8 insert-fake-data \
```

Comment:
`insert-fake-data` wasn't built with benchmarking in mind. Depending on the schema it could be that the random-data generation is the bottle neck. A better way would be to generate a json file  use `insert-json` (insert-fake-data could be used to generate the data first, then use copy to to export and re-import via insert-json)

`insert-json` will also output average/min/max and percentile values, which is more useful then the simple time based average from `insert-fake-data`